File: corvin_operator/bridges/telegram/orchestration_handler.py
"""orchestration_handler.py — Telegram handler for orchestration completion events.

Processes standardized orchestration_complete envelopes from orchestration_router.py
and formats them as Telegram messages (HTML text + optional OGG voice note).

Input: Envelope from orchestration_router.py (already in shared outbox/telegram_*.json)
Output: Formatted Telegram message via bot.send_message() + bot.send_voice()

Error handling is fail-graceful: voice note failure does not block text delivery.
"""
from __future__ import annotations

import json
import logging
import os
import sys
from pathlib import Path
from typing import Any, Optional

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

logger = logging.getLogger(__name__)


def format_orchestration_summary(payload: dict) -> str:
    """Format orchestration event into HTML-formatted Telegram message.

    Args:
        payload: Standardized envelope from orchestration_router.py

    Returns:
        HTML-formatted text string for Telegram
    """
    event_type = payload.get("event_type", "UNKNOWN")
    task_count = payload.get("task_count", 0)
    success_count = payload.get("success_count", 0)
    failed_tasks = payload.get("failed_tasks", [])

    lines = []

    # Header with emoji indicator
    if event_type == "ORCHESTRATION_COMPLETE_SUCCESS":
        lines.append("<b>✅ All Orchestration Tasks Complete</b>")
    elif event_type == "ORCHESTRATION_COMPLETE_MIXED":
        lines.append(
            f"<b>⚠️  Orchestration Complete: {success_count}/{task_count} Succeeded</b>"
        )
    else:
        lines.append(f"<b>🔔 Orchestration Event: {event_type}</b>")

    lines.append("")  # Blank line

    # Summary stats
    lines.append(f"<i>Tasks:</i> {success_count}/{task_count} completed")

    # Duration if available
    timestamp = payload.get("timestamp")
    if timestamp:
        # Calculate rough duration from timestamp
        import time
        elapsed = time.time() - timestamp
        duration_str = _format_duration(elapsed)
        lines.append(f"<i>Duration:</i> {duration_str}")

    # Failed tasks list (if any)
    if failed_tasks:
        lines.append("")
        lines.append("<b>Failed Tasks:</b>")
        for task in failed_tasks[:5]:  # Show first 5
            task_id = task.get("task_id", "unknown")
            error = task.get("error", "unknown error")
            # Escape HTML entities
            error_safe = error.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
            lines.append(f"  • <code>{task_id}</code>: {error_safe}")
        if len(failed_tasks) > 5:
            lines.append(f"  ... and {len(failed_tasks) - 5} more")

    lines.append("")  # Blank line
    lines.append("<i>All systems ready for next operation</i>")

    return "\n".join(lines)


async def handle_orchestration_complete(
    payload: dict,
    chat_id: int | str,
    bot: Any,
) -> bool:
    """Handle orchestration_complete event and send to Telegram chat.

    This is the main entry point called by the outbox poller when a message_type
    of 'orchestration_complete' is detected.

    Args:
        payload: Standardized envelope dict from orchestration_router.py
        chat_id: Telegram numeric chat_id (int or str)
        bot: telegram.Bot instance

    Returns:
        True if at least text message was sent, False if both text and voice failed
    """
    text_sent = False
    voice_sent = False

    # 1. Format and send HTML text summary
    try:
        html_text = format_orchestration_summary(payload)
        await bot.send_message(
            int(chat_id),
            html_text,
            parse_mode="HTML",
        )
        logger.info(f"orchestration: sent text summary to chat {chat_id}")
        text_sent = True

    except Exception as e:
        logger.error(
            f"orchestration: failed to send text message to {chat_id}: {e}"
        )

    # 2. Send voice note if attachment_path is provided
    voice_path = payload.get("voice_attachment_path")
    if voice_path and os.path.exists(voice_path):
        try:
            with open(voice_path, "rb") as audio_file:
                await bot.send_voice(
                    int(chat_id),
                    audio_file,
                    caption="Background tasks completion summary",
                )
                logger.info(
                    f"orchestration: sent voice note ({voice_path}) to chat {chat_id}"
                )
                voice_sent = True

        except FileNotFoundError:
            logger.warning(
                f"orchestration: voice attachment not found: {voice_path}"
            )

        except Exception as e:
            logger.warning(
                f"orchestration: voice note delivery failed (text was sent): {e}"
            )
            # Non-blocking: text was already sent, so don't fail

    if not text_sent and not voice_sent:
        logger.error(
            f"orchestration: failed to send both text and voice to {chat_id}"
        )
        return False

    return True


def _format_duration(seconds: float) -> str:
    """Format duration in human-readable form.

    Args:
        seconds: Duration in seconds

    Returns:
        Formatted duration string (e.g., "5m 23s")
    """
    if seconds < 60:
        return f"{int(seconds)}s"
    elif seconds < 3600:
        m = int(seconds // 60)
        s = int(seconds % 60)
        return f"{m}m {s}s" if s > 0 else f"{m}m"
    else:
        h = int(seconds // 3600)
        m = int((seconds % 3600) // 60)
        return f"{h}h {m}m" if m > 0 else f"{h}h"


# --- Synchronous wrapper for daemon.js integration (if needed) ---


def handle_orchestration_complete_sync(
    payload: dict,
    chat_id: int | str,
    bot: Any,
) -> bool:
    """Synchronous wrapper for handle_orchestration_complete.

    Use this if the calling context is synchronous (not async).
    Requires an event loop to be running in the background.

    Args:
        payload: Standardized envelope dict
        chat_id: Telegram chat_id
        bot: telegram.Bot instance

    Returns:
        True if message was sent, False otherwise
    """
    import asyncio

    try:
        loop = asyncio.get_event_loop()
        if loop.is_running():
            # Event loop is running in another thread; schedule as task
            task = asyncio.run_coroutine_threadsafe(
                handle_orchestration_complete(payload, chat_id, bot),
                loop,
            )
            return task.result(timeout=30)
        else:
            # Run synchronously
            return asyncio.run(
                handle_orchestration_complete(payload, chat_id, bot)
            )

    except Exception as e:
        logger.error(f"orchestration: sync wrapper failed: {e}")
        return False

File: corvin_operator/bridges/whatsapp/orchestration_handler.js
/**
 * orchestration_handler.js — WhatsApp handler for orchestration completion events.
 *
 * Routes standardized orchestration_complete envelopes from orchestration_router.py
 * to WhatsApp via Baileys. Sends text summary + optional voice attachment (OGG).
 *
 * Integration:
 *   1. orchestration_router.py writes envelope to shared outbox/
 *   2. daemon.js processOutbox() reads envelope (message_type == 'orchestration_complete')
 *   3. handleOrchestrationComplete() sends text + voice attachment
 *   4. Non-blocking: text always succeeds; voice fails gracefully
 *
 * Requirements:
 *   - Baileys client connection (waSocket)
 *   - Voice attachment path must exist (OGG format, opus codec)
 *   - Logging via console.log / console.error (matches daemon.js style)
 *
 * ADR Reference: ADR-2027 (Discord live feed + Orchestration routing)
 */

const fs = require('fs');
const path = require('path');

/**
 * Send orchestration completion summary via WhatsApp.
 *
 * @param {Object} payload - Orchestration envelope (from orchestration_router.py)
 * @param {string} payload.to - Target WhatsApp JID / chat ID
 * @param {string} payload.text - Plain-text summary (e.g., "✅ 3 tasks completed...")
 * @param {string} payload.voice_attachment_path - Path to OGG voice attachment (optional)
 * @param {string} payload.event_type - Event type (e.g., "ORCHESTRATION_COMPLETE_SUCCESS")
 * @param {number} payload.task_count - Total tasks in batch
 * @param {number} payload.success_count - Successfully completed tasks
 * @param {Array} payload.failed_tasks - List of failed task summaries
 * @param {string} chId - Chat ID (for logging/tracing)
 * @param {Object} waSocket - Baileys WhatsApp client connection
 * @param {Function} log - Logging function (e.g., console.log from daemon.js)
 * @throws {Error} Only throws on text send failure (critical); voice failure is silent
 *
 * Flow:
 *   1. Send text summary (BLOCKING — fails if send error)
 *   2. Send voice attachment if exists (NON-BLOCKING — skips silently on error)
 *   3. Return { sent: true, text: true, voice: bool }
 */
async function handleOrchestrationComplete(payload, chId, waSocket, log) {
  if (!payload || typeof payload !== 'object') {
    throw new Error('orchestration_handler: payload must be an object');
  }
  if (!payload.to) {
    throw new Error('orchestration_handler: payload.to (WhatsApp JID) is required');
  }
  if (!payload.text) {
    throw new Error('orchestration_handler: payload.text (summary) is required');
  }
  if (!waSocket) {
    throw new Error('orchestration_handler: waSocket (Baileys client) is required');
  }
  if (typeof log !== 'function') {
    log = console.log; // Fallback to console.log if no logger provided
  }

  const targetJid = payload.to;
  let voiceSent = false;
  let textSent = false;

  try {
    // ── PHASE 1: Send text summary ─────────────────────────────────────
    // This is the primary delivery; voice is enhancement. Text must succeed
    // or the whole send is considered failed.
    if (payload.text) {
      log(`orchestration: sending text summary to ${targetJid} (msg_id=${payload.msg_id || 'unknown'}, task_count=${payload.task_count})`);

      // Construct text payload for WhatsApp
      const textPayload = {
        text: payload.text,
      };

      // Optional: Add extra context if available
      if (payload.event_type) {
        textPayload._event_type = payload.event_type; // Metadata (not sent, just tracking)
      }

      // Send via Baileys
      await safeSend(waSocket, targetJid, { text: payload.text });
      textSent = true;
      log(`orchestration: text summary sent successfully`);
    }

    // ── PHASE 2: Send voice attachment (non-blocking) ─────────────────────
    // If voice file exists and voice_attachment_path is set, attempt to send.
    // On any error (file not found, send failure, timeout), log warning but do NOT throw.
    // The text summary already delivered, so we don't fail the entire send.
    if (payload.voice_attachment_path) {
      try {
        if (!fs.existsSync(payload.voice_attachment_path)) {
          log(`orchestration: voice attachment not found: ${payload.voice_attachment_path}`);
        } else {
          log(`orchestration: sending voice attachment to ${targetJid}`);

          // Read OGG file and send as audio (ptt=true for voice-note style)
          const audioBuffer = fs.readFileSync(payload.voice_attachment_path);

          await safeSend(waSocket, targetJid, {
            audio: audioBuffer,
            mimetype: 'audio/ogg; codecs=opus',
            ptt: true,  // Voice-note style (plays inline, not downloadable)
            caption: payload.voice_caption || 'Voice Summary', // Optional caption
          });

          voiceSent = true;
          log(`orchestration: voice attachment sent successfully`);
        }
      } catch (audioErr) {
        // Non-blocking: voice attachment failed, but text already sent
        log(`orchestration: voice attachment send failed (non-blocking): ${audioErr.message}`);
        // Intentionally do NOT rethrow — the text summary is the critical delivery
      }
    }

    // ── Summary ────────────────────────────────────────────────────────
    const result = {
      sent: textSent,
      text: textSent,
      voice: voiceSent,
      targetJid,
      taskCount: payload.task_count || 0,
      successCount: payload.success_count || 0,
      failedCount: (payload.failed_tasks || []).length,
    };

    log(`orchestration: completed (text=${textSent}, voice=${voiceSent})`);
    return result;

  } catch (err) {
    // Text send failed — this is critical
    log(`orchestration: CRITICAL — text send failed: ${err.message}`);
    throw err;
  }
}

/**
 * Wrapper for Baileys client.sendMessage() with retry + error handling.
 * Matches daemon.js's safeSend() semantics.
 *
 * @param {Object} waSocket - Baileys WhatsApp client
 * @param {string} jid - Target WhatsApp JID
 * @param {Object} content - Message content (text, audio, etc.)
 * @returns {Object|null} - Sent message object or null on failure
 */
async function safeSend(waSocket, jid, content) {
  try {
    const result = await waSocket.sendMessage(jid, content);
    return result;
  } catch (err) {
    throw new Error(`safeSend failed: ${err.message}`);
  }
}

/**
 * Normalize orchestration event summary for WhatsApp display.
 *
 * Converts orchestration envelope data into human-readable text summary.
 * Used when the router doesn't provide a pre-formatted text field.
 *
 * @param {Object} payload - Orchestration envelope
 * @returns {string} - Formatted summary (e.g., "✅ 3/3 tasks completed")
 */
function formatOrchestrationSummary(payload) {
  if (!payload) return '(no summary available)';

  const { event_type, task_count, success_count, failed_tasks } = payload;
  const failCount = (failed_tasks || []).length;

  // Emoji based on outcome
  const emoji = event_type === 'ORCHESTRATION_COMPLETE_SUCCESS' ? '✅' : '⚠️';

  // Basic summary
  let summary = `${emoji} ${success_count}/${task_count} tasks completed`;

  // Add failure details if any
  if (failCount > 0) {
    summary += `\n❌ ${failCount} failed`;

    // List first 3 failures (avoid flooding the message)
    const failList = (failed_tasks || []).slice(0, 3);
    for (const fail of failList) {
      if (fail.task_id && fail.error) {
        summary += `\n  • ${fail.task_id}: ${fail.error}`;
      }
    }

    if (failCount > 3) {
      summary += `\n  • ...and ${failCount - 3} more`;
    }
  }

  return summary;
}

module.exports = {
  handleOrchestrationComplete,
  formatOrchestrationSummary,
  safeSend,
};
